src/lib.rs

```diff
@@ -38,72 +35,29 @@ use project_file_walker::get_project_file_parallel_iterator;
 use searcher::get_searcher;
 use treesitter::maybe_get_query;
 
-struct MaybeInitializedCaptureIndex(AtomicU32);
-
-impl MaybeInitializedCaptureIndex {
-    const UNINITIALIZED: u32 = u32::MAX;
-    const FAILED: u32 = u32::MAX - 1;
-
-    fn mark_failed(&self) -> bool {
-        loop {
-            let existing_value = self.0.load(Ordering::Relaxed);
-            if existing_value == Self::FAILED {
-                return false;
-            }
-            let did_store = self.0.compare_exchange(
-                existing_value,
-                Self::FAILED,
-                Ordering::Relaxed,
-                Ordering::Relaxed,
-            );
-            if did_store.is_ok() {
-                return true;
-            }
-        }
-    }
+#[derive(Default)]
+struct CaptureIndex(OnceLock<Result<u32, ()>>);
 
-    pub fn get(&self) -> Result<Option<u32>, ()> {
-        let loaded = self.0.load(Ordering::Relaxed);
-        match loaded {
-            loaded if loaded == Self::UNINITIALIZED => Ok(None),
-            loaded if loaded == Self::FAILED => Err(()),
-            loaded => Ok(Some(loaded)),
-        }
-    }
-
-    pub fn get_or_initialize(&self, query: &Query, capture_name: Option<&str>) -> Result<u32, ()> {
-        if let Some(already_initialized) = self.get()? {
-            return Ok(already_initialized);
-        }
-        let capture_index = match capture_name {
-            None => 0,
-            Some(capture_name) => {
-                let capture_index = query.capture_index_for_name(capture_name);
-                if capture_index.is_none() {
-                    let did_mark_failed = self.mark_failed();
-                    if did_mark_failed {
-                        fail(&format!("invalid capture name '{}'", capture_name));
-                    } else {
-                        // whichever other thread "won the race" will have called this fail()
-                        // so we'll be getting killed shortly?
-                        thread::sleep(Duration::from_millis(100_000));
-                    }
+impl CaptureIndex {
+    pub fn get_or_init(&self, query: &Query, capture_name: Option<&str>) -> u32 {
+        let mut did_mark_failed = false;
```

Comment:
My understanding would be that if instead of "hoisting" this state like this we just directly called `fail()` in the failure case (in the inner `.get_or_init()` below) then the `OnceLock` would never actually get initialized to an `Err` (I don't know what actually happens when you call `process::exit()` from inside its `.get_or_init()` callback?) so then other "simultaneous-ish" callers wouldn't nicely see an `Err` and say "ok I'll go to sleep until I get killed"?

Reply:
Good question, I'm not even really sure in general how `process::exit` works in general if you're calling it from something other than the "main" thread of the app. Somehow it coordinates terminating all the threads, or maybe the OS just kills them?

It looks like a regular panic inside `get_or_init` propagates to the caller, but doesn't "ruin" the lock: https://doc.rust-lang.org/stable/std/sync/struct.OnceLock.html#panics

(Which is unlike mutexes, I think. Those can be "poisoned" if a panic happens in some cases, and then they can't be used any more.)

Reply:
https://doc.rust-lang.org/stable/std/sync/struct.PoisonError.html